### PR TITLE
Make an experiment fetcher

### LINKED ISSUE #29: Experiment queue fetcher for display in `iquip/apps/scheduler.py`
### Feature you want to implement
From `artiq_master`, the experiment queue is returned when `$ artiq_client show schedule`.
Here, it is again implemented in `main.py` to fetch this information and send it to `scheduler.py` to display the queued experiments.

### PR DESCRIPTION
**Pull request by giwon2004**
The function `get_experiment_queue()` returns a dictionary like:
```
{'167': {'pipeline': 'main', 'expid': {'log_level': 30, 'class_name': None, 'arguments': {'user': 'QuIQCL', 'time': 1.0, 'save': False, 'color': 'r'}, 'file': 'DIRECTORY-PATH'}, 'priority': 1, 'due_date': None, 'flush': False, 'status': 'analyzing', 'repo_msg': None}, 
 '168': {'pipeline': 'main', 'expid': {'log_level': 30, 'class_name': None, 'arguments': {'user': 'QuIQCL', 'time': 1.0, 'save': False, 'color': 'r'}, 'file': 'DIRECTORY-PATH'}, 'priority': 1, 'due_date': None, 'flush': False, 'status': 'pending', 'repo_msg': None}, 
 '169': {'pipeline': 'main', 'expid': {'log_level': 30, 'class_name': None, 'arguments': {'user': 'QuIQCL', 'time': 1.0, 'save': False, 'color': 'r'}, 'file': 'DIRECTORY-PATH'}, 'priority': 1, 'due_date': None, 'flush': False, 'status': 'pending', 'repo_msg': None}, 
 '170': {'pipeline': 'main', 'expid': {'log_level': 30, 'class_name': None, 'arguments': {'user': 'QuIQCL', 'time': 1.0, 'save': False, 'color': 'r'}, 'file': 'DIRECTORY-PATH'}, 'priority': 1, 'due_date': None, 'flush': False, 'status': 'pending', 'repo_msg': None}}
```

which is the dictionary of scheduled experiments, whose keys are the`rid`value.

This dictionary is fetched from `scheduler.py` (introduced in issue [#72](https://github.com/snu-quiqcl/iquip/issues/72) in`iquip`and pull request [#91](https://github.com/snu-quiqcl/iquip/pull/91)) to display these experiments.

This closes #29.

### CODE REVIEW

**Comment by BECATRUE** [comment]
You can use several keywords, i.e. `close` and `resolve` to close a specific issue when the PR is merged.
Please see [Linking a PR to an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#about-linked-issues-and-pull-requests).

File: main.py [hunk 1]
```diff
@@ -95,6 +95,19 @@ async def get_experiment_info(file: str) -> Any:
     return remote.examine(file)
 
 
+@app.get("/experiment/queue/", response_model=Dict)
```
Comment (BECATRUE) [inline main.py]: Please write down the type annotation in more detail.

File: main.py [hunk 2]
```diff
@@ -95,6 +95,19 @@ async def get_experiment_info(file: str) -> Any:
...
+async def get_experiment_queue() -> Dict:
+    """Gets the list of queued experiment returns it.
```
Comment (BECATRUE) [inline main.py]: Did you intend "... experiments and returns ..."?

File: main.py [hunk 3]
```diff
@@ -95,6 +95,19 @@ async def get_experiment_info(file: str) -> Any:
...
+
+    Returns:
+        A dictionary of queued experiments with rid as their keys.
```
Comment (BECATRUE) [inline main.py:109]: Please write it down in more detail.
For example, what is each value and if the value is a dictionary, then what are its keys and values.

File: main.py [hunk 4]
```diff
@@ -95,6 +95,19 @@ async def get_experiment_info(file: str) -> Any:
...
+        It includes the running experiment with different "status" value.
+    """
+    remote = get_client("master_schedule")
+    current_experiment = remote.get_status()
```
Comment (BECATRUE) [inline main.py]: Wow, how can you find this method😮

File: main.py [hunk 5]
```diff
@@ -95,6 +95,19 @@ async def get_experiment_info(file: str) -> Any:
...
+    return current_experiment
```
Comment (BECATRUE) [inline main.py]: Please return directly, not assign it to an unnecessary variable.

**Review by BECATRUE — CHANGES_REQUESTED** [review]
How does the IQUIP scheduler update its status?
Does it request to the proxy server periodically?

File: main.py (hunk 4 above)
Comment (giwon2004) [reply main.py]: By backtracking..? 😂

**Comment by giwon2004** [comment]
> How does the IQUIP scheduler update its status? Does it request to the proxy server periodically?

Yes, that is what I have made so far, but it is subject to change: see [discussion #90](https://github.com/snu-quiqcl/iquip/discussions/90) in iquip.

File: main.py (hunk 1 above)
Comment (kangz12345) [reply main.py]: E.g., `Dict[str, Any]` (please use the exact type, this is just an example).

Comment (giwon2004) [reply main.py]: Is it resolved now?

Comment (BECATRUE) [reply main.py]: Umm.. how about `Dict[int, Dict[str, Any]]`?

Comment (giwon2004) [reply main.py]: Okay, I will fix it that way.

**Comment by giwon2004** [comment]
Please check [Discussion #94](https://github.com/snu-quiqcl/iquip/discussions/94) in iquip.

File: main.py [hunk 6]
```diff
@@ -95,6 +96,28 @@ async def get_experiment_info(file: str) -> Any:
     return remote.examine(file)
 
 
+previous_queue = {}
+current_queue = {}
+@app.get("/experiment/queue/", response_model=Dict[int, Dict[str, Any]])
```
Comment (BECATRUE) [inline main.py]: Please add two empty lines after defining `current_queue`.

Comment (BECATRUE) [reply main.py]: Also, as I know, the `response_model` is only necessary when it returns a Pydantic model.
Please check it!

File: main.py [hunk 7]
```diff
@@ -95,6 +96,28 @@ async def get_experiment_info(file: str) -> Any:
...
+async def get_experiment_queue() -> Dict[int, Dict[str, Any]]:
+    """Gets the list of queued experiment and returns it.
+
+    Returns:
+        A dictionary of queued experiments with rid as their keys.
+        The value of each corresponding rid is a dictionary with several information:
+          "priority", "status", "flush", "pipeline", "expid", and etc.
```
Comment (BECATRUE) [inline main.py]: It is a minor suggestion, but please remove "and".
The "etc" includes the meaning of "and".

Comment (BECATRUE) [reply main.py]: There is the `timed` (or due date) key, right?

I think we need the information, so it looks better to include this key in the above docstring!
And it is okay to remove "fluse" as we probably won't use it.

**Review by BECATRUE — CHANGES_REQUESTED** [review]

**Review by BECATRUE — DISMISSED** [review]
LGTM!

**Review by kangz12345 — APPROVED** [review]
OK for now but we still have to consider:
1. Efficiency: does `remove.get_status()` return immediately or not? If it does, the RPC happens too frequently.
2. Timeout: I think it is more stable if a timeout is set. For example, when the queue is never changed, then it will not respond back, and the socket connection might disappear.

File: main.py [hunk 8]
```diff
@@ -95,6 +96,30 @@ async def get_experiment_info(file: str) -> Any:
     return remote.examine(file)
 
 
+previous_queue = {}
+current_queue = {}
```
Comment (kangz12345) [inline main.py]: Do we need both dictionaries? I think 1 is enough to store the previous state. The current state can be handled locally I guess.

**Review by kangz12345 — CHANGES_REQUESTED** [review]

**Comment by kangz12345** [comment]
I don't know why it has conflicts but please resolve it as well!

**Comment by giwon2004** [comment]
> OK for now but we still have to consider:
> 
> 1. Efficiency: does `remote.get_status()` return immediately or not? If it does, the RPC happens too frequently.
> 2. Timeout: I think it is more stable if a timeout is set. For example, when the queue is never changed, then it will not respond back, and the socket connection might disappear.

1. It does, but status changes like `run_done` to `analyzing` is almost at that frequency too. Though I think this calls too frequently. (That is why we use `Subscriber`, because we do not have to request again and again to track changes.)
2. Okay, I agree. I will do that in my `scheduler.py` branch.

**Comment by giwon2004** [comment]
I will resolve conflicts after PR #46 is merged.

**Comment by kangz12345** [comment]
> 1. It does, but status changes like `run_done` to `analyzing` is almost at that frequency too. Though I think this calls too frequently. (That is why we use `Subscriber`, because we do not have to request again and again to track changes.)

Exactly! I wonder why you don't use `Subscriber` here.
However, for now, I think we need to implement more features instead of optimizing the performance. Please move on to your next goal.

**Review by kangz12345 — DISMISSED** [review]
LGTM

**Comment by kangz12345** [comment]
Please resolve the conflict. I think `get_experiment_code()` should be removed again.

**Comment by giwon2004** [comment]
Done it!

**Review by kangz12345 — APPROVED** [review]

**Review by BECATRUE — APPROVED** [review]
LGTM!

**Pull request merged**